### ISSUE TITLE
v2.7.x: panic when TLS-SNI is received for an unknown domain

### ISSUE THREAD

**Issue by otbutz** [caddyserver/caddy#5680]
caddy v2.7.2 installed via the apt repository:

```
Aug 03 08:55:31 proxy caddy[1163492]: panic: runtime error: invalid memory address or nil pointer dereference
Aug 03 08:55:31 proxy caddy[1163492]: [signal SIGSEGV: segmentation violation code=0x1 addr=0x30 pc=0x903750]
Aug 03 08:55:31 proxy caddy[1163492]: goroutine 57 [running]:
Aug 03 08:55:31 proxy caddy[1163492]: github.com/caddyserver/certmagic.(*Config).getCertDuringHandshake(0xc000543520, {0x1f09a88, 0xc000194008}, _, _)
Aug 03 08:55:31 proxy caddy[1163492]:         github.com/caddyserver/certmagic@v0.19.1/handshake.go:378 +0x1390
Aug 03 08:55:31 proxy caddy[1163492]: github.com/caddyserver/certmagic.(*Config).GetCertificateWithContext(0xc000543520, {0x1f09a88, 0xc000194008}, 0xc000543450)
Aug 03 08:55:31 proxy caddy[1163492]:         github.com/caddyserver/certmagic@v0.19.1/handshake.go:84 +0xbff
Aug 03 08:55:31 proxy caddy[1163492]: github.com/caddyserver/certmagic.(*Config).GetCertificate(0xc000138ee0?, 0xc0001dc1b0?)
Aug 03 08:55:31 proxy caddy[1163492]:         github.com/caddyserver/certmagic@v0.19.1/handshake.go:50 +0x2a
Aug 03 08:55:31 proxy caddy[1163492]: github.com/caddyserver/caddy/v2/modules/caddytls.(*ConnectionPolicy).buildStandardTLSConfig.func1(0xc000543450)
Aug 03 08:55:31 proxy caddy[1163492]:         github.com/caddyserver/caddy/v2@v2.7.2/modules/caddytls/connpolicy.go:232 +0x14f
Aug 03 08:55:31 proxy caddy[1163492]: github.com/quic-go/qtls-go1-20.(*config).getCertificate(0xc0008d7800, 0xc000543450)
Aug 03 08:55:31 proxy caddy[1163492]:         github.com/quic-go/qtls-go1-20@v0.3.0/common.go:1086 +0x42
Aug 03 08:55:31 proxy caddy[1163492]: github.com/quic-go/qtls-go1-20.(*serverHandshakeStateTLS13).pickCertificate(0xc000631be8)
Aug 03 08:55:31 proxy caddy[1163492]:         github.com/quic-go/qtls-go1-20@v0.3.0/handshake_server_tls13.go:415 +0x66
Aug 03 08:55:31 proxy caddy[1163492]: github.com/quic-go/qtls-go1-20.(*serverHandshakeStateTLS13).handshake(0xc000631be8)
Aug 03 08:55:31 proxy caddy[1163492]:         github.com/quic-go/qtls-go1-20@v0.3.0/handshake_server_tls13.go:60 +0x53
Aug 03 08:55:31 proxy caddy[1163492]: github.com/quic-go/qtls-go1-20.(*Conn).serverHandshake(0xc00053d180, {0x1f09a50, 0xc000562fa0})
Aug 03 08:55:31 proxy caddy[1163492]:         github.com/quic-go/qtls-go1-20@v0.3.0/handshake_server.go:53 +0x188
Aug 03 08:55:31 proxy caddy[1163492]: github.com/quic-go/qtls-go1-20.(*Conn).handshakeContext(0xc00053d180, {0x1f09af8, 0xc0001753e0})
Aug 03 08:55:31 proxy caddy[1163492]:         github.com/quic-go/qtls-go1-20@v0.3.0/conn.go:1540 +0x3ce
Aug 03 08:55:31 proxy caddy[1163492]: github.com/quic-go/qtls-go1-20.(*Conn).HandshakeContext(0xc00005dfd0?, {0x1f09af8?, 0xc0001753e0?})
Aug 03 08:55:31 proxy caddy[1163492]:         github.com/quic-go/qtls-go1-20@v0.3.0/conn.go:1480 +0x25
Aug 03 08:55:31 proxy caddy[1163492]: created by github.com/quic-go/qtls-go1-20.(*QUICConn).Start
Aug 03 08:55:31 proxy caddy[1163492]:         github.com/quic-go/qtls-go1-20@v0.3.0/quic.go:179 +0xcf
```

Might be related to https://github.com/golang/go/issues/61639 ?

**Comment by otbutz** [comment]
@marten-seemann https://github.com/quic-go/quic-go/pull/4001

**Comment by marten-seemann** [comment]
This is where the nil pointer dereference happens: https://github.com/caddyserver/certmagic/blob/v0.19.1/handshake.go#L378

This should have been fixed by https://github.com/quic-go/quic-go/pull/4001. Did you make sure that this patch is included here (released as v0.37.1)?

**Comment by francislavoie** [comment]
Yes @marten-seemann we built and released with 0.37.1 https://github.com/caddyserver/caddy/blob/e2fc08bd34cc17b8cbb6ac364fa1ec41c4c643b9/go.mod#L20

**Comment by francislavoie** [comment]
Oh, we built with Go 1.20.6 though :thinking: did we need to build with 1.20.7?

**Comment by marten-seemann** [comment]
> Oh, we built with Go 1.20.6 though 🤔 did we need to build with 1.20.7?
> 
> Edit: Nevermind that shouldn't have mattered according to the changes in https://github.com/golang/go/issues?q=milestone%3AGo1.20.7

That's a different issue for which I'm about to cut a quic-go patch release today (the RSA key size DoS vulnerability fixed in Go 1.20.7, which I need to backport on my crypto/tls fork). I suggested including this in the Caddy patch release in https://github.com/caddyserver/caddy/issues/5671#issuecomment-1663111895.

**Comment by marten-seemann** [comment]
We should probably figure out why this panic is occurring before I cut v0.37.2. Any hints?

**Comment by animosity22** [comment]
I installed my first version from a download from the caddy website with the extra modules selected:

```
[felix@phoenix caddy]$ ./caddy build-info | head -5
go	go1.20.1
path	caddy
mod	caddy	(devel)
dep	filippo.io/edwards25519	v1.0.0	h1:0wAIcmJUqRdI8IJ/3eGi5/HwXZWPujYXXlkrQogz0Ek=
dep	github.com/AndreasBriese/bbloom	v0.0.0-20190825152654-46b345b51c96	h1:cTp8I5+VIoKjsnZuH8vjyaysT/ses3EvZeaV/1UkF2M=
```

My other box has it built from xcaddy with the latest GO version installed, which was 1.20.7.

```
[felix@gemini caddy]$ ./caddy build-info  | head -5
go	go1.20.7
path	caddy
mod	caddy	(devel)
dep	filippo.io/edwards25519	v1.0.0	h1:0wAIcmJUqRdI8IJ/3eGi5/HwXZWPujYXXlkrQogz0Ek=
dep	github.com/AndreasBriese/bbloom	v0.0.0-20190825152654-46b345b51c96	h1:cTp8I5+VIoKjsnZuH8vjyaysT/ses3EvZeaV/1UkF2M=
```

I've been up for ~10 minutes now without issues and it would panic after 2-3 minutes for me anyway.

**Comment by mholt** [comment]
@animosity22 In your build-info output, what is the version of quic-go?

**Comment by mholt** [comment]
@otbutz @animosity22 Additionally, can you provide a minimal reproducer for this? At least your config and possibly a request that triggers it?

**Comment by animosity22** [comment]
Here are the two:

```
[felix@gemini caddy]$ ./caddy build-info | grep quic
dep	github.com/quic-go/qpack	v0.4.0	h1:Cr9BXA1sQS2SmDUWjSofMPNKmvF6IiIfDRmgU0w1ZCo=
dep	github.com/quic-go/qtls-go1-20	v0.3.0	h1:NrCXmDl8BddZwO67vlvEpBTwT89bJfKYygxv4HQvuDk=
dep	github.com/quic-go/quic-go	v0.37.1	h1:M+mcsFq9KoxVjCetIwH65TvusW1UdRBc6zmxI6pkeD0=
[felix@gemini caddy]$ ./caddy version
v2.7.2 h1:QqThyoyUFAv1B7A2NMeaWlz7xmgKqU49PXBX08A+6xg=
[felix@gemini caddy]$
```

and

```
[felix@phoenix caddy]$ ./caddy build-info | head -5
go	go1.20.1
path	caddy
mod	caddy	(devel)
dep	filippo.io/edwards25519	v1.0.0	h1:0wAIcmJUqRdI8IJ/3eGi5/HwXZWPujYXXlkrQogz0Ek=
dep	github.com/AndreasBriese/bbloom	v0.0.0-20190825152654-46b345b51c96	h1:cTp8I5+VIoKjsnZuH8vjyaysT/ses3EvZeaV/1UkF2M=
[felix@phoenix caddy]$ ./caddy build-info | grep quic
dep	github.com/quic-go/qpack	v0.4.0	h1:Cr9BXA1sQS2SmDUWjSofMPNKmvF6IiIfDRmgU0w1ZCo=
dep	github.com/quic-go/qtls-go1-20	v0.1.0	h1:d1PK3ErFy9t7zxKsG3NXBJXZjp/kMLoIb3y/kV54oAI=
dep	github.com/quic-go/quic-go	v0.32.0	h1:lY02md31s1JgPiiyfqJijpu/UX/Iun304FI3yUqX7tA=
[felix@phoenix caddy]$ ./caddy version
v2.6.4 h1:2hwYqiRwk1tf3VruhMpLcYTg+11fCdr8S3jhNAdnPy8=
```

My config is small I think as I only use the security plugin to do SSO auth and that protects my site.

```
[felix@phoenix caddy]$ cat Caddyfile
{
	email {env.EMAIL}
	storage file_system {
		root /opt/caddy/ssl
	}
	order authenticate before respond
	order authorize before basicauth

	security {
		oauth identity provider github {env.GITHUB_CLIENT_ID} {env.GITHUB_CLIENT_SECRET}

		authentication portal myportal {
			crypto default token lifetime 604800
			crypto key sign-verify {env.JWT_SHARED_KEY}
			cookie domain blah.us
			cookie lifetime 604800
			enable identity provider github
			ui {
				links {
					"My Identity" "/whoami" icon "las la-user"
				}
			}

			transform user {
				match realm github
				match sub github.com/user1234
				action add role authp/admin
			}
		}

		authorization policy mypolicy {
			set auth url https://auth.blah.us/oauth2/github
			crypto key verify {env.JWT_SHARED_KEY}
			allow roles authp/admin authp/user
			validate bearer header
			inject headers with claims
		}
	}
}

auth.blah.us {
	tls {
		dns cloudflare {env.CLOUDFLARE_API_TOKEN}
		resolvers 1.1.1.1
	}
	authenticate with myportal
}
```

I wasn't doing anything specific to generate as the automatic SSL renewal would kick in and that would produce the error as it happened right after startup.

**Comment by mholt** [comment]
@animosity22 Thanks. The phoenix instance is clearly using too-old of a quic-go version for some reason: `dep	github.com/quic-go/quic-go	v0.32.0	`. It needs to be on 0.37.1 or newer.

Do they both have the same config? I wonder if a plugin is causing the downgrade.

But then again, @otbutz says he gets the error using Caddy from the apt repository... (meaning without plugins) :thinking: We are also deploying that Caddy on our website and I can't find this error in our logs at all.

**Comment by animosity22** [comment]
Oh sorry, one second, I forgot I downgraded my phoenix back back to the older version, let me update it and reproduce.

**Comment by raregems-io** [comment]
Just installed 2.7.2 on Debian 12 via this repo:
https://dl.cloudsmith.io/public/caddy/stable/deb/debian

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x30 pc=0x903750]
goroutine 28 [running]:
github.com/caddyserver/certmagic.(*Config).getCertDuringHandshake(0xc0003cf6c0, {0x1f09a88, 0xc000042058}, _, _)
        github.com/caddyserver/certmagic@v0.19.1/handshake.go:378 +0x1390
github.com/caddyserver/certmagic.(*Config).GetCertificateWithContext(0xc0003cf6c0, {0x1f09a88, 0xc000042058}, 0xc0003cf5f0)
        github.com/caddyserver/certmagic@v0.19.1/handshake.go:84 +0xbff
github.com/caddyserver/certmagic.(*Config).GetCertificate(0xc0007f2ee0?, 0xc000855620?)
        github.com/caddyserver/certmagic@v0.19.1/handshake.go:50 +0x2a
github.com/caddyserver/caddy/v2/modules/caddytls.(*ConnectionPolicy).buildStandardTLSConfig.func1(0xc0003cf5f0)
        github.com/caddyserver/caddy/v2@v2.7.2/modules/caddytls/connpolicy.go:232 +0x14f
github.com/quic-go/qtls-go1-20.(*config).getCertificate(0xc000003680, 0xc0003cf5f0)
        github.com/quic-go/qtls-go1-20@v0.3.0/common.go:1086 +0x42
github.com/quic-go/qtls-go1-20.(*serverHandshakeStateTLS13).pickCertificate(0xc000867be8)
        github.com/quic-go/qtls-go1-20@v0.3.0/handshake_server_tls13.go:415 +0x66
github.com/quic-go/qtls-go1-20.(*serverHandshakeStateTLS13).handshake(0xc000867be8)
        github.com/quic-go/qtls-go1-20@v0.3.0/handshake_server_tls13.go:60 +0x53
github.com/quic-go/qtls-go1-20.(*Conn).serverHandshake(0xc0003e7180, {0x1f09a50, 0xc00003f130})
        github.com/quic-go/qtls-go1-20@v0.3.0/handshake_server.go:53 +0x188
github.com/quic-go/qtls-go1-20.(*Conn).handshakeContext(0xc0003e7180, {0x1f09af8, 0xc0005c08d0})
        github.com/quic-go/qtls-go1-20@v0.3.0/conn.go:1540 +0x3ce
github.com/quic-go/qtls-go1-20.(*Conn).HandshakeContext(0xc00041f7d0?, {0x1f09af8?, 0xc0005c08d0?})
        github.com/quic-go/qtls-go1-20@v0.3.0/conn.go:1480 +0x25
created by github.com/quic-go/qtls-go1-20.(*QUICConn).Start
        github.com/quic-go/qtls-go1-20@v0.3.0/quic.go:179 +0xcf
```

my Caddyfile:

```
{
	email redacted@example.com
}

(cors) {
	@cors_preflight method OPTIONS
	@cors header Origin {args.0}

	handle @cors_preflight {
		header Access-Control-Allow-Origin {args.0}
		header Access-Control-Allow-Methods "GET"
		header Access-Control-Allow-Headers "X-Requested-With"
		header Access-Control-Max-Age 3600
		respond "" 204
	}

	handle @cors {
		header Access-Control-Allow-Origin {args.0}
	}
}

(global-headers) {
	header {
		-Server
		Strict-Transport-Security "max-age=31536000; includeSubDomains; preload"
		X-Content-Type-Options nosniff
		X-Frame-Options SAMEORIGIN
	}
}

(php) {
	php_fastcgi unix//run/php/php8.1-fpm.sock
}

import sites/*
```

and an example sites/site:

```
example.com {

	root * /srv/www/example/public
	file_server
	encode zstd gzip

	import global-headers

}
```

**Comment by animosity22** [comment]
Hmm, I've tried a bit more on the reproducing the issue side with a few scenarios.

- I deleted my ssl directory and let it regenerate certs on my non important box and that worked without issue
- I swapped around binaries from the website download and ones built by xcaddy and that worked without issue

@raregems-io - if you restart a few times, does it clear up?

**Comment by mholt** [comment]
Thanks for the info. I'll try to reproduce it.

I should point out that our production website and the forums are both using 2.7.2 with similar configs without issues.

**Comment by marten-seemann** [comment]
lmk if there’s anything you need from the quic-go side. I’ll be holding off on cutting the RSA patch release until we’ve reached a conclusion here.

**Comment by animosity22** [comment]
Yeah, I hate bugs like this as it happened to me twice on one box and zero times on my other box. Same build process, identical config minus 2 different site names.

```
Aug  3 07:45:49 gemini caddy[27228]: panic: runtime error: invalid memory address or nil pointer dereference
Aug  3 07:49:28 gemini caddy[27478]: panic: runtime error: invalid memory address or nil pointer dereference
```

and since those 2, I've tried restarting ~30 times and I can't get it to reproduce.

**Comment by mholt** [comment]
@marten-seemann Thanks. Really appreciate your collaboration and patience. Are there any other code paths that could allow that Conn() or RemoteAddr() to be nil? Given the spurious nature of the reports, it seems like there's either some nondeterminism here or some other code path that we didn't think of that causes them to not be populated.

**Comment by bt90** [comment]
Is that codepath always used or only if a on-demand  certificate is created? That would explain why the error isn't that common as the mechanism for HTTP2 and 1.1 isn't affected.

**Comment by mholt** [comment]
@raregems-io I am unable to reproduce the issue with that config, using HTTP/3, and trying in a loop of about a thousand requests... Tried restarting the server several times and still can't reproduce it.

**Comment by RainmakerRaw** [comment]
I seem to have this issue also, but missed this existing one (sorry!). See #5681 and merge if desired. Apologies!

**Comment by RainmakerRaw** [comment]
@mholt How can I help you reproduce this? It happens every single time I visit the URL without fail. Restart the caddy.service and it's fine (and serving the other sites without issue), but visit Plex or Jellyfin and boom, issue recurs again and again. I'm happy to liaise with you if you want to let me know how I can help?

**Comment by mholt** [comment]
@RainmakerRaw I too have a Jellyfin installation. I can try proxying to it? But I feel like it shouldn't matter what kind of content we're serving, since the bug occurs before connections are even established.

> e: I notice the site loads OK in Firefox, no issues. Dev tools shows some items using http 1.1 and others (over half the resources) are loaded with http3. Using Brave, the site refuses to load at all and the Caddy server crashes. I wonder if the agent you're testing from is having an impact (or lack of)?

If that's the case, then why aren't we seeing this on our production caddyserver.com and caddy.community sites?

The bug does seem sporadic though. @marten-seemann I dunno if any of this info is helpful, but I'm starting to wonder if there's another code path we missed. I'm not too familiar with the quic-go library though to know.

**Comment by RainmakerRaw** [comment]
@mholt Aha! I don't know what effect this has, but I can bypass the issue and make Brave work again. On my test desktop client, Firefox was using DoH by default (my own AdGuard Home installation on a VPS). Brave was set to use the local DNS server, and causes Caddy to crash. When I set Brave to also use offsite DoH the Jellyfin server loads fine again and Caddy doesn't crash. I'll be honest, I don't know what this means, but it's progress maybe?

**Comment by RainmakerRaw** [comment]
Further digging: Cloudflare has enabled ECH (encrypted client hello) on my domain. The Caddy server (and Jellyfin server) are running on the same LAN as my test desktop device - i.e. I'm hairpinning back in through the router to connect. 

The local DNS server and the remote DoH server are both AdGuard Home using the same encrypted upstreams, but I wonder if something's getting mangled and causing the crash? I can connect to Plex and Jellyfin (using their domain not the local address) perfectly now I have enabled DoH in the browser. Both browsers are set to use ECH where available - is yours? 

In summary, ECH is enabled on both browsers. With DoH and ECH enabled, Firefox connects to Jellyfin fine and Caddy doesn't crash. With Brave, ECH being enabled but DoH not causes Caddy to crash and no page loads. Adding DoH to the ECH on Brave makes everything work OK again.

**Comment by marten-seemann** [comment]
> @marten-seemann I dunno if any of this info is helpful, but I'm starting to wonder if there's another code path we missed. I'm not too familiar with the quic-go library though to know.

Good thinking! In fact there is: `tls.Config.GetCertificate` also passes in a `ClientHelloInfo`. I feel really stupid for missing this 😔. For some reason I thought that `tls.Config.GetConfigForClient` would be the only place, and didn't even bother checking. This means we'll need to wrap `GetCertificate` as well, equivalent to what we do in https://github.com/quic-go/quic-go/pull/4001/files#diff-c1eebd1c52f66da524b99c57f22669500f22784ae80475fbf2b786bc0d1ba278.

Fix coming later today. This will be included in the v0.37.2 patch release.

**Comment by RainmakerRaw** [comment]
@marten-seemann This ties in with my last edit above. ECH was enabled in Brave for me, and I needed to also enable DoH to not have Caddy crash. Enabling ECH + DoH meant no crash and the page loads. Disabling ECH in the browser means the page loads (and Caddy doesn't crash) regardless of DoH status.

**Comment by mholt** [comment]
@marten-seemann Oh, you are awesome. Thank you so much. I don't know if I would have been able to figure that out. We're really lucky to have you building and maintaining that lib!

Once we merge the fix I would like as many people involved here to confirm that it works if you could please. :sweat_smile:

**Comment by marten-seemann** [comment]
Fix is ready at https://github.com/quic-go/quic-go/pull/4014.

**Comment by mholt** [comment]
You're a wizard Marten. Thanks so much for the patch!

**Comment by mholt** [comment]
@otbutz @bt90 @animosity22 @raregems-io @RainmakerRaw Since I was never able to reproduce this, could you all try it out? I've pushed a commit that upgrades quic-go with the patch to master, so if you build caddy from source or use xcaddy like `xcaddy build master ...` you should be able to verify whether this works.

As soon as we get some confirmation, I'll tag One. More. Release.

Thank you!

**Comment by RainmakerRaw** [comment]
@mholt Sorry Matt it's 2am here, I only just saw this by chance. Compiling now...

**Comment by RainmakerRaw** [comment]
@mholt bad news: 

```
caddy version
v2.7.3-0.20230804004403-51b1bfb12505 h1:bOvmG2P3zkUCfK+uPZUFTnTYmdI3MZ3fDSKms0jgWtc=
```

```
sudo journalctl -xeu caddy
░░
░░ The job identifier is 6612.
Aug 04 02:32:54 rock5b caddy[11574]: panic: runtime error: invalid memory address or nil pointer dereference
Aug 04 02:32:54 rock5b caddy[11574]: [signal SIGSEGV: segmentation violation code=0x1 addr=0x30 pc=0x47b1dc]
Aug 04 02:32:54 rock5b caddy[11574]: goroutine 167 [running]:
Aug 04 02:32:54 rock5b caddy[11574]: github.com/caddyserver/certmagic.(*Config).getCertDuringHandshake(0x400086c820, {0x1963238, 0x40001a6008}, _, _)
Aug 04 02:32:54 rock5b caddy[11574]:         github.com/caddyserver/certmagic@v0.19.1/handshake.go:378 +0xf8c
Aug 04 02:32:54 rock5b caddy[11574]: github.com/caddyserver/certmagic.(*Config).GetCertificateWithContext(0x400086c820, {0x1963238, 0x40001a6008}, 0x400086c7>
Aug 04 02:32:54 rock5b caddy[11574]:         github.com/caddyserver/certmagic@v0.19.1/handshake.go:84 +0x9c4
Aug 04 02:32:54 rock5b caddy[11574]: github.com/caddyserver/certmagic.(*Config).GetCertificate(0x40009fe380?, 0x4000734738?)
Aug 04 02:32:54 rock5b caddy[11574]:         github.com/caddyserver/certmagic@v0.19.1/handshake.go:50 +0x30
Aug 04 02:32:54 rock5b caddy[11574]: github.com/caddyserver/caddy/v2/modules/caddytls.(*ConnectionPolicy).buildStandardTLSConfig.func1(0x400086c750)
Aug 04 02:32:54 rock5b caddy[11574]:         github.com/caddyserver/caddy/v2@v2.7.3-0.20230804004403-51b1bfb12505/modules/caddytls/connpolicy.go:232 +0x144
Aug 04 02:32:54 rock5b caddy[11574]: github.com/quic-go/qtls-go1-20.(*config).getCertificate(0x4000780600, 0x400086c750)
Aug 04 02:32:54 rock5b caddy[11574]:         github.com/quic-go/qtls-go1-20@v0.3.1/common.go:1086 +0x44
Aug 04 02:32:54 rock5b caddy[11574]: github.com/quic-go/qtls-go1-20.(*serverHandshakeStateTLS13).pickCertificate(0x4000621bb0)
Aug 04 02:32:54 rock5b caddy[11574]:         github.com/quic-go/qtls-go1-20@v0.3.1/handshake_server_tls13.go:415 +0x5c
Aug 04 02:32:54 rock5b caddy[11574]: github.com/quic-go/qtls-go1-20.(*serverHandshakeStateTLS13).handshake(0x4000621bb0)
Aug 04 02:32:54 rock5b caddy[11574]:         github.com/quic-go/qtls-go1-20@v0.3.1/handshake_server_tls13.go:60 +0x40
Aug 04 02:32:54 rock5b caddy[11574]: github.com/quic-go/qtls-go1-20.(*Conn).serverHandshake(0x40007a1180, {0x1963200, 0x40002ae460})
Aug 04 02:32:54 rock5b caddy[11574]:         github.com/quic-go/qtls-go1-20@v0.3.1/handshake_server.go:53 +0x138
Aug 04 02:32:54 rock5b caddy[11574]: github.com/quic-go/qtls-go1-20.(*Conn).handshakeContext(0x40007a1180, {0x19632a8, 0x400001bcb0})
Aug 04 02:32:54 rock5b caddy[11574]:         github.com/quic-go/qtls-go1-20@v0.3.1/conn.go:1540 +0x348
Aug 04 02:32:54 rock5b caddy[11574]: github.com/quic-go/qtls-go1-20.(*Conn).HandshakeContext(0x0?, {0x19632a8?, 0x400001bcb0?})
Aug 04 02:32:54 rock5b caddy[11574]:         github.com/quic-go/qtls-go1-20@v0.3.1/conn.go:1480 +0x24
Aug 04 02:32:54 rock5b caddy[11574]: created by github.com/quic-go/qtls-go1-20.(*QUICConn).Start
Aug 04 02:32:54 rock5b caddy[11574]:         github.com/quic-go/qtls-go1-20@v0.3.1/quic.go:179 +0xc8
Aug 04 02:32:54 rock5b systemd[1]: caddy.service: Main process exited, code=exited, status=2/INVALIDARGUMENT
░░ Subject: Unit process exited
░░ Defined-By: systemd
░░ Support: https://www.debian.org/support
░░
░░ An ExecStart= process belonging to unit caddy.service has exited.
░░
░░ The process' exit code is 'exited' and its exit status is 2.
Aug 04 02:32:54 rock5b systemd[1]: caddy.service: Failed with result 'exit-code'.
░░ Subject: Unit failed
░░ Defined-By: systemd
░░ Support: https://www.debian.org/support
░░
░░ The unit caddy.service has entered the 'failed' state with result 'exit-code'.
```

**Comment by mholt** [comment]
@marten-seemann Dang, there's still a code path we're missing.

@RainmakerRaw Thank you so much for trying it out. Sorry to keep you up :sweat_smile: Get some good sleep, hopefully we'll have it figured out soon.

**Comment by RainmakerRaw** [comment]
Reverting to v2.6.4 in the interim. You'll get there guys, thanks for all your hard work!

**Comment by marten-seemann** [comment]
> @marten-seemann Dang, there's still a code path we're missing.

Where? I don't see it. There's only the two code paths that I fixed: https://pkg.go.dev/crypto/tls (search for `ClientHelloInfo`).

**Comment by RainmakerRaw** [comment]
@mholt @marten-seemann Guys I didn't go to sleep (yet)... I got 2.7.3 working by changing a setting in Cloudflare. Just about to test 2.7.2 also just in case. I clearly am missing something here. To stop this being more of an XY problem: 

My domain is registered at Cloudflare. I have a free CF account and so can't proxy Plex, Jellyfin etc per ToS. DNS at Cloudflare is set like this: 

```
A domain.com 1.2.3.4 # My home WAN IP, proxied by Cloudflare)
CNAME jellyfin.domain.com domain.com (**not** proxied by Cloudflare as it's video)
```

Cloudflare has enabled ECH on my domain. Firefox and Brave are both set up to use ECH where available, and both are using DoH to my own off-prem server (AdGuard Home on a VPS). When connecting to jellyfin.domain.com in Firefox, everything works. When connecting using Brave, Caddy crashes. 

**However**, I have now discovered that setting the CNAME for Jellyfin (and Plex) to temporarily also be proxied by Cloudflare (no video served, just to test the home page loading) then everything works, even in Brave! No crash on Caddy. 

Is this helpful?

**Comment by marten-seemann** [comment]
The Go standard library doesn't even support ECH. It seems like debugging thing 2 layers higher than is useful, I think it would be more helpful to figure out which code path is taken that leads to the panic.

**Comment by RainmakerRaw** [comment]
@marten-seemann I see. My sincere apologies then. I just noticed that I can make Caddy crash or work by enabling and disabling ECH in the browser, or by toggling proxying in CF. As it turns out, changing Jellyfin's and Plex's CNAME for an A record in CF has now stopped Caddy crashing at all, so far. Everything works OK in all tested browsers and v2.7.3 doesn't crash any more. I guess it's a coincidence, you certainly would know more than me, I'm just a (new) user with a enough *nix knowledge to be dangerous haha.

>  I think it would be more helpful to figure out which code path is taken that leads to the panic.

In that case I'm definitely of no use to you. Sorry for cluttering up the thread (and pinging you unnecessarily). Best of luck solving it, I'll keep an eye out for updates. Best wishes.

**Comment by mholt** [comment]
It's weird that some deployments don't see it at all (like ours), but others ONLY see this. I can't figure it out.

**Comment by francislavoie** [comment]
Okay! I have a way to replicate it.

Simple config:

```
{
    debug
    admin off
    http_port 8881
}

https://localhost:8882 {
    respond "foo"
}
```

(I'm using non-standard ports because I have things on 80/443 on my machine already)

Making a request for the _wrong SNI_ (i.e. `foo.localhost` when Caddy only has `localhost`) with this command (container with a custom curl with http3 support, because it's the easiest way for me to try http3 reliably):

```
$ docker run -it --rm --net=host ymuski/curl-http3 curl -vsk -o/dev/null --http3-only https://foo.localhost:8882
```

@marten-seemann hopefully that should be enough for you to be able to do some `fmt.Printf()` tracing and such?

**Comment by mholt** [comment]
@marten-seemann If you're staying up late because of this, don't :innocent: -- Francis and I think it might be related to some code in Caddy now. https://github.com/caddyserver/caddy/blob/master/listeners.go#L511-L530

**Comment by francislavoie** [comment]
I have a possible patch (seems to work!) but it doesn't feel correct to me and I don't know if there's terrible side-effects:

```diff
diff --git a/modules/caddytls/connpolicy.go b/modules/caddytls/connpolicy.go
index a4dc4119..524fa36e 100644
--- a/modules/caddytls/connpolicy.go
+++ b/modules/caddytls/connpolicy.go
@@ -115,7 +115,13 @@ func (cp ConnectionPolicies) TLSConfig(_ caddy.Context) *tls.Config {
                                                continue policyLoop
                                        }
                                }
-                               return pol.TLSConfig, nil
+                               cfg := pol.TLSConfig.Clone()
+                               gc := cfg.GetCertificate
+                               cfg.GetCertificate = func(clientHello *tls.ClientHelloInfo) (*tls.Certificate, error) {
+                                       clientHello.Conn = hello.Conn
+                                       return gc(clientHello)
+                               }
+                               return cfg, nil
                        }
 
                        return nil, fmt.Errorf("no server TLS configuration available for ClientHello: %+v", hello)
```

**Comment by marten-seemann** [comment]
@francislavoie Have you figured out why it's panicking without this change?

**Comment by francislavoie** [comment]
Not quite.

Basically, we have the concept of "connection policies" (i.e. Caddy config which chooses a TLS config to use depending on SNI or client certs, etc) and quic-go doesn't see those underlying TLS configs. Our `GetConfigForClient` returns one of those configs.

The `GetConfigForClient` _does_ have the correct `hello.Conn`, so we can wrap it ourselves and pass it down and that seems to fix it.

But what I _don't_ know is _why_ the underlying `GetCertificate` gets a "wrong" hello (with the conn missing). What else is creating/managing/cloning a hello? I don't think we do anything like that in Caddy, so I feel like that must be something in quic-go or stdlib.

**Comment by marten-seemann** [comment]
Are you saying that `GetConfigForClient` might return a `tls.Config`, and then the `GetCertificate` callback is called on that config? Does quic-go need to handle this recursive case?

**Comment by francislavoie** [comment]
> Are you saying that `GetConfigForClient` might return a `tls.Config`

Yes; that _is_ the point of that callback! :sweat_smile: It returns a config that is appropriate given the `hello` (i.e. different config per SNI)

> and then the `GetCertificate` callback is called on that config?

Yeah, but I don't know who calls this. I think stdlib does, according to the stacktrace, with `crypto/tls.(*serverHandshakeStateTLS13).pickCertificate`

> Does quic-go need to handle this recursive case?

I really don't know whose job it is to handle this. I don't know where the `hello` struct is coming from.

**Comment by mholt** [comment]
I'm off to sleep for the night, but in the morning I may try a git bisect to see where this started happening and see if that helps us pinpoint it some more.

But yeah, `GetConfigForClient` is a case where the initial TLS config is only used to get a different config based on the properties of the handshake.

However, I don't believe it would be called more than once per handshake :thinking:

**Comment by WeidiDeng** [comment]
quic-go can't handle this recursive call without forking stdlib. I can modify caddy to handle this situation, but the problem is the `nil` net.Conn.

**Comment by WeidiDeng** [comment]
@otbutz @animosity22 @raregems-io  Can you try `xcaddy build fix-quic-nil-conn` to see if it's fixed?

**Comment by xet7** [comment]
@WeidiDeng

Thanks! I compiled Caddy with your PR. Now Caddy does not crash anymore at my server !

**Comment by animosity22** [comment]
Minus my first two panics, I can't reproduce the issue anymore. I've tried a number of patterns with no luck.

I only use Chrome and no other browser on my home devices.

I do use Cloudflare. I tried toggling the proxy on and off for a few services. I tried putting a bogus DNS entry and hitting that Caddy and that would not reproduce it.

I'm stumped on whatever caused mine unless it was dumb luck of someone probing me at the same time but I have not seen the issue surface since.

**Comment by xet7** [comment]
I use CloudFlare for nameserver and TLS certs. After that PR, there is no crashes. I did not change anything in Caddyfile or CloudFlare settings. Only added that PR, then Caddy did not crash anymore.

**Comment by WeidiDeng** [comment]
@W0n9 Does it always happen?

**Comment by francislavoie** [comment]
Wait what, @W0n9 that's a different error entirely 😱

@marten-seemann that one seems to be entirely in quic-go. You might need to look into that 😔

**Comment by francislavoie** [comment]
That reads to me like that one is likely when quic-go is writing out the response. That's very different than what we were seeing before. 😓

**Comment by wazerstar** [comment]
Oh yeah got this today on build 2.7.2 from the day it was released..


```
{"level":"debug","ts":1691155488.2340865,"logger":"http.stdlib","msg":"http: TLS handshake error from 162.158.254.74:65009: EOF"}
{"level":"debug","ts":1691155488.3216188,"logger":"http.stdlib","msg":"http: TLS handshake error from 162.158.254.74:52411: EOF"}
{"level":"debug","ts":1691155488.4074762,"logger":"http.stdlib","msg":"http: TLS handshake error from 162.158.254.74:65533: EOF"}
{"level":"debug","ts":1691155488.4976387,"logger":"http.stdlib","msg":"http: TLS handshake error from 162.158.254.74:13325: EOF"}
{"level":"debug","ts":1691155488.5845616,"logger":"http.stdlib","msg":"http: TLS handshake error from 162.158.254.74:54355: EOF"}
{"level":"debug","ts":1691155510.6032279,"logger":"http.stdlib","msg":"http: TLS handshake error from 172.68.38.70:25147: EOF"}
{"level":"debug","ts":1691155510.7825186,"logger":"http.stdlib","msg":"http: TLS handshake error from 172.68.38.70:53343: EOF"}
{"level":"debug","ts":1691155510.978358,"logger":"http.stdlib","msg":"http: TLS handshake error from 172.68.38.70:27969: EOF"}
{"level":"debug","ts":1691155511.1574907,"logger":"http.stdlib","msg":"http: TLS handshake error from 172.68.38.70:31599: EOF"}
{"level":"debug","ts":1691155511.3371205,"logger":"http.stdlib","msg":"http: TLS handshake error from 172.68.38.70:34977: EOF"}
{"level":"debug","ts":1691155564.7425537,"logger":"http.stdlib","msg":"http: TLS handshake error from 162.158.194.72:56943: EOF"}
{"level":"debug","ts":1691155564.9266932,"logger":"http.stdlib","msg":"http: TLS handshake error from 162.158.194.72:57421: EOF"}
{"level":"debug","ts":1691155565.0714908,"logger":"http.stdlib","msg":"http: TLS handshake error from 162.158.194.72:51845: EOF"}
{"level":"debug","ts":1691155565.212855,"logger":"http.stdlib","msg":"http: TLS handshake error from 162.158.194.72:36793: EOF"}
{"level":"debug","ts":1691155565.3699675,"logger":"http.stdlib","msg":"http: TLS handshake error from 162.158.194.72:33257: EOF"}
{"level":"debug","ts":1691155590.0132987,"logger":"http.stdlib","msg":"http: TLS handshake error from 172.69.245.5:12355: EOF"}
{"level":"debug","ts":1691155590.0775383,"logger":"http.stdlib","msg":"http: TLS handshake error from 172.69.76.147:43931: EOF"}
{"level":"debug","ts":1691155590.1332562,"logger":"http.stdlib","msg":"http: TLS handshake error from 172.69.245.5:14631: EOF"}
{"level":"debug","ts":1691155590.2817445,"logger":"http.stdlib","msg":"http: TLS handshake error from 172.69.245.5:65205: EOF"}
{"level":"debug","ts":1691155590.2939334,"logger":"http.stdlib","msg":"http: TLS handshake error from 172.69.76.147:20245: EOF"}
{"level":"debug","ts":1691155590.425773,"logger":"http.stdlib","msg":"http: TLS handshake error from 172.69.245.5:24625: EOF"}
{"level":"debug","ts":1691155590.487499,"logger":"http.stdlib","msg":"http: TLS handshake error from 172.69.76.147:25683: EOF"}
{"level":"debug","ts":1691155590.5614686,"logger":"http.stdlib","msg":"http: TLS handshake error from 172.69.245.5:31177: EOF"}
{"level":"debug","ts":1691155590.6855319,"logger":"http.stdlib","msg":"http: TLS handshake error from 172.69.76.147:22955: EOF"}
{"level":"debug","ts":1691155590.88261,"logger":"http.stdlib","msg":"http: TLS handshake error from 172.69.76.147:41557: EOF"}
```

**Comment by marten-seemann** [comment]
I'll try to get a fix out for the recursive `GetConfigForClient` / `GetCertificate` some time tonight.

**Comment by mholt** [comment]
Thanks Marten. Appreciate your help :100:

The other panic that is being reported just above comes from this code:

https://github.com/quic-go/quic-go/blob/18d3846d4f96f24f62bd8af2339fb860bb420be6/internal/ackhandler/sent_packet_handler.go#L294-L303

(`pnSpace.largestSent` --  apparently `pnSpace` is nil)

**Comment by marten-seemann** [comment]
Hi guys, can you please test with https://github.com/quic-go/quic-go/pull/4016. I'd like to get confirmation that this actually works before cutting yet another patch release again.

**Comment by francislavoie** [comment]
Thanks @marten-seemann, I can confirm my reproduce case from https://github.com/caddyserver/caddy/issues/5680#issuecomment-1664889820 is fixed by https://github.com/quic-go/quic-go/pull/4016.

**Comment by RainmakerRaw** [comment]
@mholt Sorry, per my last reply I fixed the crash (not the underlying cause, just the manifestation of it) by changing my DNS and ECH settings on Cloudflare. I had my root domain proxied by CF but the subdomains in question were CNAMES yet not proxied. This seems to have caused issues with DNS (tagging my Jellyfin server with an ECH https answer that wasn't valid). 

Once I removed the CF proxy from the root domain, DNS answers changed to reflect the CNAME and root domain, and Caddy no longer crashed. I can replicate the original conditions and wait for DNS to propagate and caches to expire (I chain from DHCP > local DNS > my upstream DNS) to test if you wish; it'll just take me some time. I had already built the binary with that patch, and put it in a dir on the server just in case...

**Comment by mholt** [comment]
@RainmakerRaw A confirmation would be helpful. The changes you made are unrelated -- even if correct -- except that they just happen to make it so that the clients send a known/expected SNI to Caddy, which is the real reason it "fixed" it for you. The bug is exposed when the SNI is not known and Caddy doesn't have a cert. You don't have to revert your changes necessarily, just send a request to Caddy with an unknown SNI/hostname.

**Comment by RainmakerRaw** [comment]
@mholt Ah, I understand what you were asking now. Following the earlier example with Docker and curl-http3 (substituting for `justdanz/curl-http3` because I'm on arm64), I tried with Caddy v2.7.2 as a control and it did indeed crash. 

I then tried with `GOOS=linux GOARCH=arm64 ./xcaddy build --with github.com/caddy-dns/cloudflare --with github.com/quic-go/quic-go=github.com/quic-go/quic-go@280054c` and there is no crash after several minutes of running. I hope this is helpful.

**Comment by mholt** [comment]
Awesome, thank you. Confirmations of the fix from anyone else experiencing the issue would be valuable as we await the patch release! :100:

**Comment by AlyoshaVasilieva** [comment]
Testing my server with `curl -v https://google.com --connect-to google.com:443:MY_SERVER_IP:443 --http3-only`

Caddy crash: 2.7.2 built with `xcaddy build --with github.com/caddyserver/cache-handler --with github.com/caddyserver/replace-response`  
Works: 2.7.2 built with `xcaddy build --with github.com/caddyserver/cache-handler --with github.com/caddyserver/replace-response --with github.com/quic-go/quic-go=github.com/quic-go/quic-go@280054c`

**Comment by animosity22** [comment]
I did a quick test as well:

```
docker run -it --rm ymuski/curl-http3 curl -v https://google.com --connect-to google.com:443:192.168.1.25:443 --http3-only
```
2.7.2 fails but the updated build works.

**Comment by mholt** [comment]
Excellent, thank you both!

@marten-seemann We now have numerous confirmations of the patch working, so I'm confident with your fix. :100: Thank you thank you.

**Comment by wazerstar** [comment]
For whatever its worth, I tested on windows as well and I no longer receive TLS handshake error spam after building with

`xcaddy build master --with github.com/caddy-dns/cloudflare --with github.com/caddyserver/transform-encoder --with github.com/WeidiDeng/caddy-cloudflare-ip --with github.com/quic-go/quic-go=github.com/quic-go/quic-go@280054c`

**Comment by wazerstar** [comment]
Oh, as I just checked again I got handshake error now?

{"level":"debug","ts":1691191817.8557942,"logger":"http.stdlib","msg":"http: TLS handshake error from 192.168.1.26:56616: no certificate available for '192.168.1.4'"}

**Comment by bt90** [comment]
You don't have a certificate for the raw IP, so that is expected and only logged with level debug.

**Comment by wazerstar** [comment]
> You don't have a certificate for the raw IP, so that is expected and only logged with level debug.

Right, I should not stay up this late, was just focusing on watching the errors, thats from the server and not cloudflare, so all good.

**Comment by W0n9** [comment]
> Thanks @marten-seemann, I can confirm my reproduce case from [#5680 (comment)](https://github.com/caddyserver/caddy/issues/5680#issuecomment-1664889820) is fixed by [quic-go/quic-go#4016](https://github.com/quic-go/quic-go/pull/4016).
> 
> I think the panic @W0n9 reported is a separate issue though. I don't know how to replicate that problem. A new issue should probably be opened for that.



> @RainmakerRaw @otbutz @animosity22 @raregems-io @bt90 Can some of you also confirm that the patch at [quic-go/quic-go#4016](https://github.com/quic-go/quic-go/pull/4016) works for you? You will need to build Caddy with that version of quic-go. Here's an easy way using xcaddy: `xcaddy build --with github.com/quic-go/quic-go=github.com/quic-go/quic-go@280054c` -- or you can simply modify the `go.mod` of Caddy to refer to that commit (`go get github.com/quic-go/quic-go@280054c`) if you're cloning the source manually.

I have tried this new commit, and panic again. I think I shoud open a new issue.😥
```
xcaddy build    --with github.com/caddy-dns/duckdns     --with github.com/W0n9/caddy_waf_plugin   --with github.com/quic-go/quic-go=github.com/quic-go/quic-go@280054c   --output /root/caddy
```
```
8月 05 10:24:36 NewCaddy caddy[1905066]: panic: runtime error: invalid memory address or nil pointer dereference
8月 05 10:24:36 NewCaddy caddy[1905066]: [signal SIGSEGV: segmentation violation code=0x1 addr=0x50 pc=0xa6532a]
8月 05 10:24:36 NewCaddy caddy[1905066]: goroutine 31817 [running]:
8月 05 10:24:36 NewCaddy caddy[1905066]: github.com/quic-go/quic-go/internal/ackhandler.(*sentPacketHandler).ReceivedAck(0xc003a69600?, 0xc003a69600?, 0x5?, {0xc000fe9802?, 0x10200002c?, 0x2b87520?})
8月 05 10:24:36 NewCaddy caddy[1905066]:         github.com/quic-go/quic-go@v0.37.1/internal/ackhandler/sent_packet_handler.go:298 +0x6a
8月 05 10:24:36 NewCaddy caddy[1905066]: github.com/quic-go/quic-go.(*connection).handleAckFrame(0xc000fe9800, 0xc003a69640, 0x60?)
8月 05 10:24:36 NewCaddy caddy[1905066]:         github.com/quic-go/quic-go@v0.37.1/connection.go:1484 +0x4e
8月 05 10:24:36 NewCaddy caddy[1905066]: github.com/quic-go/quic-go.(*connection).handleFrame(0xc000fe9800, {0x1f11860?, 0xc003a69640?}, 0x0?, {{0xd1, 0x1c, 0x94, 0x5, 0x0, 0x0, ...}, ...})
8月 05 10:24:36 NewCaddy caddy[1905066]:         github.com/quic-go/quic-go@v0.37.1/connection.go:1280 +0xed
8月 05 10:24:36 NewCaddy caddy[1905066]: github.com/quic-go/quic-go.(*connection).handleFrames(0xc000fe9800, {0xc00169520e?, 0xc000e69860?, 0x40deca?}, {{0xd1, 0x1c, 0x94, 0x5, 0x0, 0x0, ...}, ...}, ...)
8月 05 10:24:36 NewCaddy caddy[1905066]:         github.com/quic-go/quic-go@v0.37.1/connection.go:1253 +0x365
8月 05 10:24:36 NewCaddy caddy[1905066]: github.com/quic-go/quic-go.(*connection).handleUnpackedLongHeaderPacket(0xc000fe9800, 0xc001cae780, 0xa7?, {0x60?, 0x0?, 0x2b87520?}, 0x4a)
8月 05 10:24:36 NewCaddy caddy[1905066]:         github.com/quic-go/quic-go@v0.37.1/connection.go:1194 +0x6aa
8月 05 10:24:36 NewCaddy caddy[1905066]: github.com/quic-go/quic-go.(*connection).handleLongHeaderPacket(0xc000fe9800, {0xc0023da880, {0x1f130e8, 0xc001cae690}, {0xc12b8a990dc0ca26, 0x1e4bfa1fc8, 0x2b87520}, {0xc001695200, 0x4a, 0x5ac}, ...}, ...)
8月 05 10:24:36 NewCaddy caddy[1905066]:         github.com/quic-go/quic-go@v0.37.1/connection.go:967 +0x765
8月 05 10:24:36 NewCaddy caddy[1905066]: github.com/quic-go/quic-go.(*connection).handlePacketImpl(0xc000fe9800, {0xc0023da880, {0x1f130e8, 0xc001cae690}, {0xc12b8a990dc0ca26, 0x1e4bfa1fc8, 0x2b87520}, {0xc001695200, 0x4a, 0x5ac}, ...})
8月 05 10:24:36 NewCaddy caddy[1905066]:         github.com/quic-go/quic-go@v0.37.1/connection.go:849 +0x1a8
8月 05 10:24:36 NewCaddy caddy[1905066]: github.com/quic-go/quic-go.(*connection).run(0xc000fe9800)
8月 05 10:24:36 NewCaddy caddy[1905066]:         github.com/quic-go/quic-go@v0.37.1/connection.go:560 +0x525
8月 05 10:24:36 NewCaddy caddy[1905066]: created by github.com/quic-go/quic-go.(*baseServer).handleInitialImpl
8月 05 10:24:36 NewCaddy caddy[1905066]:         github.com/quic-go/quic-go@v0.37.1/server.go:672 +0x7cd
8月 05 10:24:36 NewCaddy systemd[1]: caddy.service: Main process exited, code=exited, status=2/INVALIDARGUMENT
8月 05 10:24:36 NewCaddy systemd[1]: caddy.service: Failed with result 'exit-code'.
```

**Comment by marten-seemann** [comment]
@W0n9 A fix is in the works, and will be included in the v0.37.2 release.

**Comment by francislavoie** [comment]
@W0n9 can you try a build with https://github.com/quic-go/quic-go/pull/4018 ? You can use `--with github.com/quic-go/quic-go=github.com/quic-go/quic-go@71ce36b`

**Comment by W0n9** [comment]
> @W0n9 can you try a build with [quic-go/quic-go#4018](https://github.com/quic-go/quic-go/pull/4018) ? You can use `--with github.com/quic-go/quic-go=github.com/quic-go/quic-go@ack-after-handshake-complete`

I have tried caddy 2.7.3, it fixed for me👍

**Comment by wazerstar** [comment]
Hey @mholt , @marten-seemann I'm not kidding its back with me on windows

I built master after https://github.com/caddyserver/caddy/commit/a8cc5d1a7d9d80e12ab895beaf4eaed34cc6bded

v2.7.3-0.20230805213002-65e33fc1ee47 h1:6PFhIFWiV7Nfez0pPrk8lnQ0IdbMPnnQUz1clDX+HxY=


Is it me being wrong here?, getting massive spams.

```

{"level":"debug","ts":1691309078.0863955,"logger":"http.stdlib","msg":"http: TLS handshake error from 162.158.29.26:57591: EOF"}
{"level":"debug","ts":1691309078.1616666,"logger":"http.stdlib","msg":"http: TLS handshake error from 172.71.13.71:57891: EOF"}
{"level":"debug","ts":1691309078.233539,"logger":"http.stdlib","msg":"http: TLS handshake error from 162.158.29.26:13931: EOF"}
{"level":"debug","ts":1691309078.3741865,"logger":"http.stdlib","msg":"http: TLS handshake error from 162.158.29.26:37805: EOF"}
{"level":"debug","ts":1691309078.3912568,"logger":"http.stdlib","msg":"http: TLS handshake error from 172.71.13.71:65391: EOF"}
{"level":"debug","ts":1691309078.5111217,"logger":"http.stdlib","msg":"http: TLS handshake error from 162.158.29.26:46309: EOF"}
{"level":"debug","ts":1691309078.621165,"logger":"http.stdlib","msg":"http: TLS handshake error from 172.71.13.71:24263: EOF"}
{"level":"debug","ts":1691309078.661057,"logger":"http.stdlib","msg":"http: TLS handshake error from 162.158.29.26:16085: EOF"}
{"level":"debug","ts":1691309078.8506064,"logger":"http.stdlib","msg":"http: TLS handshake error from 172.71.13.71:9705: EOF"}
{"level":"debug","ts":1691309079.0801218,"logger":"http.stdlib","msg":"http: TLS handshake error from 172.71.13.71:47291: EOF"}
{"level":"debug","ts":1691309477.3321996,"logger":"http.stdlib","msg":"http: TLS handshake error from 162.158.228.70:29757: EOF"}
{"level":"debug","ts":1691309477.637015,"logger":"http.stdlib","msg":"http: TLS handshake error from 162.158.228.70:37959: EOF"}
{"level":"debug","ts":1691309478.848605,"logger":"http.stdlib","msg":"http: TLS handshake error from 162.158.228.70:50391: EOF"}
{"level":"debug","ts":1691309479.9939373,"logger":"http.stdlib","msg":"http: TLS handshake error from 162.158.228.70:41453: EOF"}
{"level":"debug","ts":1691310489.8182952,"logger":"http.stdlib","msg":"http: TLS handshake error from 172.68.88.96:64407: EOF"}
{"level":"debug","ts":1691310489.950913,"logger":"http.stdlib","msg":"http: TLS handshake error from 172.68.88.96:42685: EOF"}
{"level":"debug","ts":1691310490.0826983,"logger":"http.stdlib","msg":"http: TLS handshake error from 172.68.88.96:13429: EOF"}
{"level":"debug","ts":1691310490.21494,"logger":"http.stdlib","msg":"http: TLS handshake error from 172.68.88.96:43121: EOF"}
{"level":"debug","ts":1691310490.3484063,"logger":"http.stdlib","msg":"http: TLS handshake error from 172.68.88.96:28693: EOF"}
```

**Comment by francislavoie** [comment]
That's normal. That's not a bug. That's just bots/crawlers trying to connect to your server without knowing the correct domain name (i.e. trying to connect by IP and no domain). Ignore those. Turn off the `debug` logging level to quiet it.

**Comment by wazerstar** [comment]
> @wazerstar That's normal. That's not a bug. That's just bots/crawlers trying to connect to your server without knowing the correct domain name (i.e. trying to connect by IP and no domain). Ignore those. Turn off the `debug` logging level to quiet it.

That's odd, why do I not receive those error when I revert back to the version previous to this, the handshake error EOF only recently started to appear after this?

I will let this run for another 12 hours and see what happens.

**Comment by francislavoie** [comment]
It's definitely not a new error, handshake errors have always been there. If you only enabled debug logging recently then you'd only be seeing those now. It's also affected by external factors whether bots are actually hitting your server or not. It happens in bursts as bots change their attention from one IP address to another.

**Comment by wazerstar** [comment]
> It's definitely not a new error, handshake errors have always been there. If you only enabled debug logging recently then you'd only be seeing those now. It's also affected by external factors whether bots are actually hitting your server or not. It happens in bursts as bots change their attention from one IP address to another.

No I have debug enabled for about 5-6 days now to figure out some other things, only after this I noticed the errors, perhaps I have been lucky not to get targeted, I don't know, sorry for causing spikes in this thread again then, will just leave and ignore it until I'm done finding cause of other issue.

**Comment by bt90** [comment]
[Baader–Meinhof phenomenon](https://en.wikipedia.org/wiki/Frequency_illusion) :wink:

**Comment by bt90** [comment]
[Baader–Meinhof phenomenon](https://en.wikipedia.org/wiki/Frequency_illusion) :wink:
